sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/SubscriptionPartitionProcessorImpl.java

```diff
@@ -110,29 +132,41 @@ private void onMessages(List<SequencedMessage> messages) {
   }
 
   @Override
-  public void failed(State from, Throwable failure) {
-    completionFuture.setException(ExtractStatus.toCanonical(failure));
+  @SuppressWarnings("argument.type.incompatible")
+  public ProcessContinuation runFor(Duration duration) {
+    Instant deadline = Instant.now().plus(duration);
+    start();
+    try (SubscriptionPartitionProcessorImpl closeThis = this) {
+      while (!completionFuture.isDone() && deadline.isAfterNow()) {
```

Comment:
This busy waiting seems inefficient, consider doing something similar to this [CancellableQueue](https://github.com/apache/beam/blob/9a934b7c62f5dee31cdb9a65ca3a79e83996281c/sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/CancellableQueue.java#L34) (or re-using it) so you don't have to use a completable future for "doneness"/"error" passing and it won't require spinning.